### PR TITLE
Cache for tables that are reused frequently

### PR DESCRIPTION
**Pull request by kderme**
Supersedes https://github.com/input-output-hk/cardano-db-sync/pull/1067

### CODE REVIEW

File: cardano-db-sync/src/Cardano/DbSync/Era/Byron/Insert.hs [hunk 1]
```diff
@@ -58,37 +60,41 @@ data ValueFee = ValueFee
...
-  -- times (ie 10x or more).
-  when (getSyncStatus details == SyncFollowing)
-    DB.transactionCommit
-  pure res
+insertByronBlock env firstBlockOfEpoch blk details = do
+    res <- runExceptT $
+              case byronBlockRaw blk of
+                Byron.ABOBBlock ablk -> insertABlock tracer cache firstBlockOfEpoch ablk details
+                Byron.ABOBBoundary abblk -> insertABOBBoundary tracer cache abblk details
+    -- Serializiing things during syncing can drastically slow down full sync
```
Comment (tmcgilchrist) [inline cardano-db-sync/src/Cardano/DbSync/Era/Byron/Insert.hs:70]:
```suggestion
    -- Serializing things during syncing can drastically slow down full sync
```

File: cardano-db-sync/src/Cardano/DbSync/Era/Shelley/Genesis.hs [hunk 2]
```diff
@@ -20,6 +20,7 @@ import           Control.Monad.Trans.Except.Extra (newExceptT)
 
 import qualified Cardano.Db as DB
 
+import           Cardano.DbSync.Cache
```
Comment (tmcgilchrist) [inline cardano-db-sync/src/Cardano/DbSync/Era/Shelley/Genesis.hs:23]:
```suggestion
import           Cardano.DbSync.Cache (Cache(..), uninitiatedCache)
```

File: cardano-db-sync/src/Cardano/DbSync/Era/Shelley/Insert.hs [hunk 3]
```diff
@@ -170,6 +174,13 @@ insertShelleyBlock tracer lenv firstBlockOfEpoch blk lStateSnap details = do
         SyncFollowing -> 10
         SyncLagging -> 2000
 
+    rightToJust (Right a) = Just a
+    rightToJust _ = Nothing
```
Comment (tmcgilchrist) [inline cardano-db-sync/src/Cardano/DbSync/Era/Shelley/Insert.hs:178]: There's a version of this in `https://hackage.haskell.org/package/either` that's already floating around in package deps somewhere.

File: cardano-db-sync/src/Cardano/DbSync/Rollback.hs [hunk 4]
```diff
@@ -60,11 +71,14 @@ rollbackToPoint backend trce point =
...
-    queryBlockId :: MonadIO m => Point CardanoBlock -> ReaderT SqlBackend m (Either DB.LookupFail DB.BlockId)
-    queryBlockId pnt =
+    queryBlock :: MonadIO m => Point CardanoBlock
+               -> ReaderT SqlBackend m (Either DB.LookupFail (DB.BlockId, Maybe Word64))
+    queryBlock pnt = do
       case getPoint pnt of
-        Origin -> DB.queryGenesis
-        At blk -> DB.queryBlockId (SBS.fromShort . getOneEraHash $ blockPointHash blk)
+        Origin ->
+          fmap (, Nothing) <$> DB.queryGenesis
```
Comment (tmcgilchrist) [inline cardano-db-sync/src/Cardano/DbSync/Rollback.hs:79]:
```suggestion
          (, Nothing) <<$>> DB.queryGenesis
```
I think there is `fmap . fmap` defined in Protolude as `<<$>>`

**Comment by tmcgilchrist** [comment]
What kind of speedup and memory usage are you seeing before / after this change?

**Comment by kderme** [comment]
> What kind of speedup and memory usage are you seeing before / after this change?

On benchmarks I've seen some pretty good speedup by caching mul https://github.com/input-output-hk/cardano-db-sync/pull/1067

**Comment by erikd** [comment]
Superceeded by PR #1094 .

**Pull request closed without merging**